### PR TITLE
Bst

### PR DESCRIPTION
**Pull request by derekmhewitt**

### CODE REVIEW

File: src/bst.py [hunk 1]
```diff
@@ -0,0 +1,285 @@
...
+        except AttributeError:
+            md = 0
+        return max(ld, rd, md)
+
+    def balance(self, starting_point=None):
+        """Will return an integer that's positive or negative that represents
+        the difference between depth on both sides from the root."""
+        if starting_point is None:
+            starting_point = self.root
+        return starting_point.left - starting_point.right
```
Comment (WillWeatherford) [inline src/bst.py]: I don't think this line is doing what you think it does. `.left` and `.right` are Node objects here, and they don't have `__sub__` methods to allow you to minus one from the other. Make sure to test your balance method.

File: src/bst.py [hunk 2]
```diff
@@ -0,0 +1,285 @@
...
+        if root:
+            self.root = Node(root)
+            self.length = 1
+        else:
+            self.root = root
+
+    def insert(self, val, data=None):
+        """Inserts a value  into the BST.  If the value is already in
+        the BST it will be ingored."""
+        if not self.contains(val):
```
Comment (WillWeatherford) [inline src/bst.py:193]: This will work; however you are substantially increasing the time it takes to run `insert`, since you are first traversing the tree to see if the item is contained, then traversing the tree again to insert the item. You can improve on this and do it all in one run through.

File: src/bst.py [hunk 3]
```diff
@@ -0,0 +1,285 @@
...
+                self.root._insert(val, data)
+            else:
+                self.root = Node(val, data)
+            self.length += 1
+
+    def contains(self, val):
+        """Will return True if val is in the BST, or False if it's not."""
+        try:
+            return self._contains(val, self.root)
+        except AttributeError:
```
Comment (WillWeatherford) [inline src/bst.py:209]: Where are you expecting the `AttributeError` to pop up? Doesn't look it can happen inside this function. If you're expecting an AttributeError to occur in `self._contains()`, your `try except` block should also be in `self._contains()`.

Comment (WillWeatherford) [reply src/bst.py:209]: Also, with this `AttributeError` passing, will `contains()` actually return False?

File: src/bst.py [hunk 4]
```diff
@@ -0,0 +1,285 @@
...
+        return self.length
+
+    def depth(self, starting_point=None):
+        """Will return the depth of the tree by counting "levels".  An empty
+        BST will return 0, a BST with 1 value will return 1, a BST with 2
+        values will return 2, and the value will fluctuate from then on
+        depending on the branch layout."""
+        if starting_point is None:
+            starting_point = self.root
+        # import pdb; pdb.set_trace();
```
Comment (WillWeatherford) [inline src/bst.py:242]: Purge commented out code from your submissions.

File: src/bst.py [hunk 5]
```diff
@@ -0,0 +1,285 @@
...
+        return starting_point.left - starting_point.right
+
+    def in_order(self, starting_point=None):
+        """
+        This function will return a generator that will return the values
+        of the tree using in-order traversal, one value at a time.
+        """
+        if starting_point is None:
+            starting_point = self.root
+        return starting_point._in_order()
```
Comment (WillWeatherford) [inline src/bst.py:275]: What's going to happen on this line and rhymes with "Cat Tribute Terror"?

Hey did you hear about the lawyer who passed the bar exam then failed horribly when representing the guitarist from U2?

HE DIDN'T TEST FOR EDGE CASES.

Comment (WillWeatherford) [reply src/bst.py:275]: Every time you write a data structure without testing every method of that data structure on an empty case, you will be subject to my bad jokes.

You have been warned.

File: src/bst.py [hunk 6]
```diff
@@ -0,0 +1,285 @@
...
+            current = unvisited.shift()
+            if current not in visited:
+                visited.append(current)
+                if current.left:
+                    unvisited.push(current.left)
+                if current.right:
+                    unvisited.push(current.right)
+                yield current.val
+
+    def delete_node(self, val):
```
Comment (WillWeatherford) [inline src/bst.py]: This should simply be named "delete", not "delete_node".

Also, need a docstring,

File: src/bst.py [hunk 7]
```diff
@@ -0,0 +1,285 @@
...
+        delete_me = self.find_node(val)
+        if delete_me is False:
+            return "That node is not in the BST."
+        left_childs = delete_me.in_order(delete_me.left)
```
Comment (WillWeatherford) [inline src/bst.py]: Whoops, that's not quite right. `delete_me` is a node, and doesn't have that attribute.

File: src/bst.py [hunk 8]
```diff
@@ -0,0 +1,464 @@
...
+        try:
+            z = pivot_parent.parent
+        except AttributeError:
+            z = None
+            self.root = b
+        try:
+            w = pivot_parent.right.left
+        except AttributeError:
+            w = None
+        b.left = a
```
Comment (WillWeatherford) [inline src/bst.py:390]: Looks like you need to have another AttributeError check here.
